test_public_network.py

```diff
@@ -112,25 +115,60 @@ def test_public_network_port_security(approach, two_servers_in_same_subnet):
     attacker.assert_run('sudo apt update --allow-releaseinfo-change')
     attacker.assert_run('sudo apt install -y curl ettercap-text-only tcpdump')
 
-    # The attacker starts poisoning the environment
-    mitm = approach['mitm']
-
-    v4 = victim.ip('public', 4)
-    v6 = victim.ip('public', 6)
+    ipv4 = victim.ip('public', 4)
+    ipv6 = victim.ip('public', 6)
+    # Get first IPv6 link local address
+    ll_addr_v6 = ip_interface(victim.output_of(
+        "ip -6 -brief addr show scope link | awk '{ print $3 }' | head -1"
+    )).ip
+    ipv4_gateway = victim.gateway('public', 4)
+    ipv6_gateway = victim.gateway('public', 6)
+    if ip_version == 4:
+        victim_ip = ipv4
+        curl_params='-4'
+        mtm_method = 'arp'
+    elif ip_version == 6:
+        victim_ip = ipv6
+        curl_params='-6'
+        mtm_method = 'ndp'
+    else:
+        raise NotImplementedError(f"Unknown IP version {ip_version}")
+
+    # This needs to match on IPv4 and IPv6 otherwise // means all addresses
+    # and will try to intercept traffic between all hosts on the network
+    # Additionally this has to match all addresses used: the IPv6 link-local
+    # address and all IPv4 gateway addresses
+    ettercap_params = oneliner(f'''
+        -M {mtm_method}:oneway
+        '/{ipv4}/{ipv6};{ll_addr_v6}/'
+        '/{ipv4_gateway};{ipv4_gateway+1};{ipv4_gateway+2}/{ipv6_gateway}/'
+    ''')
 
-    attacker.assert_run(f'sudo ettercap -D -w pcap -M {mitm} /{v4}/{v6}/')
+    # The attacker starts poisoning the environment
+    attacker.assert_run(oneliner(f"""
+        sudo systemd-run --same-dir --unit ettercap
+        ettercap -T -w pcap {ettercap_params}
+    """))
 
     # The victim initiates a slow download
     region = victim.zone['slug'][:-1]
     file = f"https://{region}-fixtures.objects.{region}.cloudscale.ch/10mib"
 
     # Download with 1MiB per second, abort if it takes longer than 15 seconds
-    victim.assert_run(f'curl {file} -4 --limit-rate 1M --max-time 15 -O')
-
-    # No TCP packets on port 443 should have been intercepted
-    intercepted_packets = int(attacker.output_of(oneliner("""
-        sudo tcpdump -r pcap tcp port 443
-        | grep -v truncated
+    # Don't fail if curl exits with a timeout (exit code 28)
```

Comment:
What does exit code 7 do?

Reply:
I improved the comment. It means "failed to connect".